### ISSUE TITLE
CS1666 when creating a span from a stack-only buffer

### ISSUE THREAD

**Issue by koszeggy** [dotnet/roslyn#57583]
**Reproduction:**

```cs
private unsafe ref struct MyRefStruct
{
    private readonly BinaryWriter writer;
    private int bufferLength;

    // we are in a ref struct so this never goes on the heap
    private fixed byte buffer[255];

    private void Dump()
    {
        Debug.Assert(bufferLength is > 0 and <= 255);
        writer.Write((byte)bufferLength);

        // CS1666: You cannot use fixed size buffers contained in unfixed expressions.
        //writer.Write(new ReadOnlySpan<byte>(buffer, bufferLength));

        // However, this works without a fixed context, even in a non-ref struct
        for (int i = 0; i < bufferLength; i++)
            writer.Write(buffer[i]);

        bufferLength = 0;
    }

    // other members here
}
```

**Expected Behavior**:

Being in a `ref struct`, the fixed buffer never gets relocated so creating the span should work.

**Remark**:

Maybe related: #39632

**Comment by jaredpar** [comment]
Closing as a duplicate of https://github.com/dotnet/csharplang/issues/1792

The compiler is operating to the language spec today. We need to change the language design first before we can change the compiler to take advantage of this behavior.
